assert/assertions.go

```diff
@@ -165,17 +165,40 @@ func ObjectsAreEqualValues(expected, actual interface{}) bool {
 		return true
 	}
 
-	actualType := reflect.TypeOf(actual)
-	if actualType == nil {
+	expectedValue := reflect.ValueOf(expected)
+	actualValue := reflect.ValueOf(actual)
+	if !expectedValue.IsValid() || !actualValue.IsValid() {
 		return false
 	}
-	expectedValue := reflect.ValueOf(expected)
-	if expectedValue.IsValid() && expectedValue.Type().ConvertibleTo(actualType) {
+
+	expectedType := expectedValue.Type()
+	actualType := actualValue.Type()
+	if !expectedType.ConvertibleTo(actualType) {
+		return false
+	}
+
+	if !isNumericType(expectedType) || !isNumericType(actualType) {
 		// Attempt comparison after type conversion
-		return reflect.DeepEqual(expectedValue.Convert(actualType).Interface(), actual)
+		return reflect.DeepEqual(
+			expectedValue.Convert(actualType).Interface(), actual,
+		)
 	}
 
-	return false
+	// If BOTH values are numeric, there are chances of false positives due
+	// to overflow or underflow. So, we need to make sure to always convert
+	// the smaller type to a larger type before comparing.
+	if expectedType.Size() >= actualType.Size() {
+		return actualValue.Convert(expectedType).Interface() == expected
+	}
+
+	return expectedValue.Convert(actualType).Interface() == actual
+}
+
+// isNumericType returns true if the type is one of:
+// int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64,
+// float32, float64
+func isNumericType(t reflect.Type) bool {
+	return t.Kind() >= reflect.Int && t.Kind() <= reflect.Float64
```

Comment:
Aha
```suggestion
	return t.Kind() >= reflect.Int && t.Kind() <= reflect.Complex128
```

Reply:
I don't think `complex` numbers should be included in this bucket. Because We will not be able to convert an `int*` `float*` to `complex*`. So, there is no basis for comparison. 

Also, this condition will make sure that _`int*/float* <> complex*`_ comparisons will always yield `false`
https://github.com/stretchr/testify/blob/2d27e0f92acbd4ef9063399060e58ba3a6867310/assert/assertions.go#L176-L178

---

I tinkered around in the playground to verify this: https://go.dev/play/p/bhIO--yWxMN

Reply:
But you can convert a complex128 to a complex64. Your test on L176 already fixes your concern by implicitly dividing the bucket into two. These testcases all pass with my proposed change:
```go
		{complex128(1e+100 + 1e+100i), complex64(complex(math.Inf(0), math.Inf(0))), false},
		{complex64(complex(math.Inf(0), math.Inf(0))), complex128(1e+100 + 1e+100i), false},
		{complex128(1e+100 + 1e+100i), 270, false},
		{270, complex128(1e+100 + 1e+100i), false},
		{complex128(1e+100 + 1e+100i), 3.14, false},
		{3.14, complex128(1e+100 + 1e+100i), false},
		{complex128(1e+10 + 1e+10i), complex64(1e+10 + 1e+10i), true},
		{complex64(1e+10 + 1e+10i), complex128(1e+10 + 1e+10i), true},
```

The interesting part is that 10+0i is mathematically equal to 10. But Go's equality operator [will not perform this operation](https://go.dev/play/p/AMYnvdCfYe1) and the types are truly [not convertible](https://go.dev/play/p/6iqAi5A7r-h) without using the helper functions. So I think it's safe and acceptable to include the behaviour in the above testcases.

Can you show me a testcase which behaves incorrectly?

NB: Careful with untyped literals when experimenting here, they [look like they are convertible](https://go.dev/play/p/5-wSmJcEFqG), but that's a static conversion so we will never see it inside ObjectsAreEqualValues.

Reply:
> But you can convert a complex128 to a complex64

Oh! thats right 🤦🏽‍♂️